pkg/services/statefetcher/statefetcher.go

```diff
@@ -352,7 +381,7 @@ func (s *Service) run() {
 			if len(batch) == 0 {
 				continue
 			}
-			if err = s.chain.AddContractStorageItems(batch, syncHeight, expectedRoot); err != nil {
+			if err = s.chain.AddContractStorageItems(batch, syncHeight, expectedRoot, witness); err != nil {
```

Comment:
I'm not sure why you're passing it always when processing batches of KVs. Intermediate roots will never be correct, only the final one can be, so to me you should be:
 * checking stateroot witness once before performing any state changes
 * adding sets of KVs as usual
 * checking the resulting stateroot to be equal to the one we proved to be correct at the first step